src/norm/mongodb.nim

```diff
@@ -0,0 +1,727 @@
+# | --------- | ------------- |---------------------
+# | float     | Double        | default float in Nim already 64-bit
+# | string    | String (UTF8) | string already UTF-8
+# | Oid       | Oid           | https://nim-lang.org/docs/oids.html; all-zeroes count as a MISSING entry; not a null
+# | bool      | Bool          | 
+# | timestamp | TimeUTC       | https://nim-lang.org/docs/times.html; 1970-01-01 00:00:00 counts as MISSING; not null
+# | Option[T] | Null          | Only fields of Option[T] can be null
+# | int       | Int32         |
+#
+# NOT YET SUPPORTED (compile-time error generated):
+#
+# | float32   | Double        | 
+# | float64   | Double        | default float in Nim already 64-bit
+# | (Object)  | Document      | it is not possible to a "Missing" state. The closest analogue is Option[ObjectType]
+# | seq[T]    | Array         | to support null, make it seq[Option[T]]
+# | int32     | Int32         |
+# | int64     | Int64         |
+# Tuples
+# Tables
+#
+
+import
+  strutils,
+  sequtils,
+  macros,
+  typetraits,
+  logging,
+  options,
+  tables,
+  oids,
+  times
+
+import
+  nimongo/bson,
+  nimongo/mongo
+
+import
+  rowutils,
+  objutils,
+  pragmas
+
+
+export
+  strutils,
+  sequtils,
+  macros,
+  logging,
+  options,
+  tables,
+  oids,
+  times
+
+export
+  bson,
+  mongo
+
+export
+  rowutils,
+  objutils,
+  pragmas
+
+# compile-time globals
+var
+  normObjectNamesRegistry {.compileTime.}: seq[string] = @[]
+
+const NORM_UNIVERSAL_TYPE_LIST* = @[
+  "float",
+  "string",
+  "Oid",
+  "bool",
+  "Time",
+  "int",
+  "Option[float]",
+  "Option[string]",
+  "Option[Oid]",
+  "Option[bool]",
+  "Option[Time]",
+  "Option[int]",
+  "seq[float]",
+  "seq[string]",
+  "seq[Oid]",
+  "seq[bool]",
+  "seq[Time]",
+  "seq[int]"
+]
+
+proc getCollectionName*(objRepr: ObjRepr): string =
+  ##[ Get the name of the DB table for the given object representation:
+  ``table`` pragma value if it exists or lowercased type name otherwise.
+  ]##
+
+  result = objRepr.signature.name.toLowerAscii()
+
+  for prag in objRepr.signature.pragmas:
+    if prag.name == "table" and prag.kind == pkKval:
+      return $prag.value
+
+
+proc getCollectionName*(T: typedesc): string =
+  ##[ Get the name of the DB table for the given type: ``table`` pragma value if it exists
+  or lowercased type name otherwise.
+  ]##
+
+  when T.hasCustomPragma(table): T.getCustomPragmaVal(table)
+  else: ($T).toLowerAscii()  
+
+
+var
+  allCollections*: seq[string] = @[]
+  normConnection*: string
+  normUser*: string
+  normPassword*: string
+  normDatabase*: string
+
+template genWithoutDb(newCollections: seq[string]): untyped {.dirty.} =
+  ## simply add to global allCollections
+
+  allCollections.insert(newCollections)
+
+template genWithDb(connection, user, password, database: string, useParams: int, newCollections: seq[string]): untyped {.dirty.} =
+  ## Generate ``withDb`` templates.
+
+  allCollections.insert(newCollections)
+  if useParams == -1:

```

Comment:
The title implies it's a bool flag (use of don't use params). But it's not, and it's not immediately clear what it means. The number of params? But why can it be -1?

Reply:
Also, this -1 number looks like a magic number out of nowhere.

Reply:
For reasons I don't fully understand, Nim turns template bool parameters into ints dynamically, breaking the code. Very frustrating. I would prefer it to be a bool. Anyway, a -1 is all binary ones, so I'm using it to represent True and a zero as false. I'll add a comment to explain that. I might also submit a bug report for Nim.

Reply:
Turns out to be a known bug: https://forum.nim-lang.org/t/4995

Reply:
Looks like there is a PR for Nim, but it is not in the release candidate. It might be a while before we can use bool template parameters safely.